### ISSUE TITLE
feature request: save animations as GIFs

### ISSUE THREAD

**Issue by jeremymanning** [ContextLab/hypertools#120]
in addition to support for saving animations as .avi files, also support saving as .gif animations.

**Comment by rarredon** [comment]
Hypertools should currently support saving animations to AVI files. The kw arg `save_path` just has to have the `.avi` postfix. This is because `ffmpeg`, the tool currently used to output the animations, supports writing to AVI files. To see what other file formats are supported by `ffmpeg`, you can run `ffmpeg -formats` in the command line.

Now onto exporting as GIF. The `ffmpeg` tool does not support writing to GIF, however, the `imagemagick` tool does. So I tried changing the code

```python
    # save
    if save_path is not None:
        if animate:
            Writer = animation.writers['ffmpeg']
            writer = Writer(fps=frame_rate, bitrate=1800)
            line_ani.save(save_path, writer=writer)
```

to

```python
    # save
    if save_path is not None:
        if animate:
            if str.lower(save_path).endswith('.gif'):
                if 'imagemagick' in animation.writers.list():
                    Writer = animation.writers['imagemagick']
                else:
                    warnings.warn('Must install ImageMagick to export GIFs. '
                                  'Proceding without exporting animation.')
            else:
                Writer = animation.writers['ffmpeg']
            writer = Writer(fps=frame_rate, bitrate=1800)
            line_ani.save(save_path, writer=writer)
```

and then I ran the script in `examples/save_movie.py` but with `save_path='animation.gif'`. When I run this I get the error `ValueError: Invalid file object: <_io.BufferedReader name=7>` buried deep underneath a large stack of function calls. I can't even begin to decipher this error. I would think there's an issue in my code, or that exporting to GIF doesn't work, however, I was able to successfully export to GIF by following the basic example [here](https://matplotlib.org/2.0.0/examples/animation/basic_example.html) (with minor changes).

Now I noticed that when I run the script in `examples/save_movie.py`, the memory usage on my computer shoots up. So I think that the exception being triggered is somehow related to my computer running out of RAM. My computer has 4 GB of RAM (not much for modern computing). I am curious if I would get the same error on a computer with 8 GB or 16 GB of RAM. 

**TL;DR:** I suggest implementing the changes above and running the `examples/save_movie.py` script on a computer with >4 GB of RAM. I hope this helps!

**Comment by andrewheusser** [comment]
thanks @rarredon! i'll give that a try

**Comment by qihongl** [comment]
Hi, are people still working on this issue? 

And currently, what's the best way of saving an animation plot?

Thanks!

**Comment by jeremymanning** [comment]
@qihongl did you install `ffmpeg`?  That's needed to save animations...

**Comment by qihongl** [comment]
Thanks! @jeremymanning  
Yes! What's the list of supported extensions for animations?

**Comment by jeremymanning** [comment]
Everything is done through `ffmpeg`-- so whatever formats are supported in your install of `ffmpeg` should be supported by `hypertools`.  To get a list of formats (in Terminal), type `ffmpeg -formats` and to get a list of supported codecs, type `ffmpeg -codecs`.

**Comment by qihongl** [comment]
Thanks again! @jeremymanning 

Hm... Then I guess maybe the error message is more complicated than I thought. 

When I set `animate = False`, I can save it (e.g. as `.png`), but when I set `animate = True`, I can't save it in any format (error message attached below), such as `.gif`, although `.gif` is listed in `ffmpeg -formats`. Has you seen this error before? I can save the geo object if anyone want to replicate this error. 

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-18-774c23f7fe6c> in <module>()
      9     title = title_text,
     10     save_path='temppic.gif',
---> 11     size = (6, 5)
     12 )

~/.conda/envs/brainiak/lib/python3.6/site-packages/hypertools/plot/plot.py in plot(x, fmt, marker, markers, linestyle, linestyles, color, colors, palette, group, hue, labels, legend, title, size, elev, azim, ndims, model, model_params, reduce, cluster, align, normalize, n_clusters, save_path, animate, duration, tail_duration, rotations, zoom, chemtrails, precog, bullettime, frame_rate, explore, show, transform, vectorizer, semantic, corpus, ax)
    431             Writer = animation.writers['ffmpeg']
    432             writer = Writer(fps=frame_rate, bitrate=1800)
--> 433             line_ani.save(save_path, writer=writer)
    434 
    435         else:

~/.conda/envs/brainiak/lib/python3.6/site-packages/matplotlib/animation.py in save(self, filename, writer, fps, dpi, codec, bitrate, extra_args, metadata, extra_anim, savefig_kwargs)
   1258                         # TODO: See if turning off blit is really necessary
   1259                         anim._draw_next_frame(d, blit=False)
-> 1260                     writer.grab_frame(**savefig_kwargs)
   1261 
   1262         # Reconnect signal for first draw if necessary

~/.conda/envs/brainiak/lib/python3.6/contextlib.py in __exit__(self, type, value, traceback)
     97                 value = type()
     98             try:
---> 99                 self.gen.throw(type, value, traceback)
    100             except StopIteration as exc:
    101                 # Suppress StopIteration *unless* it's the same exception that

~/.conda/envs/brainiak/lib/python3.6/site-packages/matplotlib/animation.py in saving(self, fig, outfile, dpi, *args, **kwargs)
    235             yield self
    236         finally:
--> 237             self.finish()
    238 
    239 

~/.conda/envs/brainiak/lib/python3.6/site-packages/matplotlib/animation.py in finish(self)
    367     def finish(self):
    368         '''Finish any processing for writing the movie.'''
--> 369         self.cleanup()
    370 
    371     def grab_frame(self, **savefig_kwargs):

~/.conda/envs/brainiak/lib/python3.6/site-packages/matplotlib/animation.py in cleanup(self)
    406     def cleanup(self):
    407         '''Clean-up and collect the process used to write the movie file.'''
--> 408         out, err = self._proc.communicate()
    409         self._frame_sink().close()
    410         verbose.report('MovieWriter -- '

~/.conda/envs/brainiak/lib/python3.6/subprocess.py in communicate(self, input, timeout)
    841 
    842             try:
--> 843                 stdout, stderr = self._communicate(input, endtime, timeout)
    844             finally:
    845                 self._communication_started = True

~/.conda/envs/brainiak/lib/python3.6/subprocess.py in _communicate(self, input, endtime, orig_timeout)
   1503                     selector.register(self.stdin, selectors.EVENT_WRITE)
   1504                 if self.stdout:
-> 1505                     selector.register(self.stdout, selectors.EVENT_READ)
   1506                 if self.stderr:
   1507                     selector.register(self.stderr, selectors.EVENT_READ)

~/.conda/envs/brainiak/lib/python3.6/selectors.py in register(self, fileobj, events, data)
    349 
    350         def register(self, fileobj, events, data=None):
--> 351             key = super().register(fileobj, events, data)
    352             poll_events = 0
    353             if events & EVENT_READ:

~/.conda/envs/brainiak/lib/python3.6/selectors.py in register(self, fileobj, events, data)
    235             raise ValueError("Invalid events: {!r}".format(events))
    236 
--> 237         key = SelectorKey(fileobj, self._fileobj_lookup(fileobj), events, data)
    238 
    239         if key.fd in self._fd_to_key:

~/.conda/envs/brainiak/lib/python3.6/selectors.py in _fileobj_lookup(self, fileobj)
    222         """
    223         try:
--> 224             return _fileobj_to_fd(fileobj)
    225         except ValueError:
    226             # Do an exhaustive search.

~/.conda/envs/brainiak/lib/python3.6/selectors.py in _fileobj_to_fd(fileobj)
     37         except (AttributeError, TypeError, ValueError):
     38             raise ValueError("Invalid file object: "
---> 39                              "{!r}".format(fileobj)) from None
     40     if fd < 0:
     41         raise ValueError("Invalid file descriptor: {}".format(fd))

ValueError: Invalid file object: <_io.BufferedReader name=54>
```

**Comment by andrewheusser** [comment]
hey @qihongl - could you save to mp4? e.g.: https://github.com/ContextLab/hypertools/blob/master/examples/save_movie.py

**Comment by qihongl** [comment]
Cool! It works! Thank you so much! @andrewheusser @jeremymanning 
I still prefer `.gif` though, is this currently supported?

**Comment by andrewheusser** [comment]
It's on the list, but hasn't been implemented yet.  You are welcome to give it a shot if you'd be up for it!  

Since it's working with mp4, you can easily save it as a gif with ffmpeg outside of hypertools: https://superuser.com/questions/556029/how-do-i-convert-a-video-to-gif-using-ffmpeg-with-reasonable-quality

**Comment by qihongl** [comment]
@andrewheusser Thank you so much! 👍 I might try that!

**Comment by jeremymanning** [comment]
This is addressed in the HyperTools 1.0 update (PR #272, https://github.com/ContextLab/hypertools/pull/272). `hyp.plot(..., animate=True, save_path=...)` now dispatches on the file extension and uses matplotlib's `PillowWriter` for `.gif` (and `.png`/`.apng`), so GIF export works with no ffmpeg/ImageMagick dependency and without the `ValueError: Invalid file object` crash reported in this thread.

```python
import numpy as np, hypertools as hyp
data = [np.cumsum(np.random.randn(60, 3), axis=0)]
hyp.plot(data, animate=True, save_path='animation.gif', frame_rate=20, show=False)
```
produces a real 640x480 GIF (confirmed via `file animation.gif` -> `GIF image data, version 89a`).

Closing as addressed in the 1.0 update -- please reopen if you hit anything else.
